src/solver/tron.jl

```diff
@@ -0,0 +1,403 @@
+
+"""`slope, qs = compute_Hs_slope_qs!(Hs, H, s, g)`
+
+Computes
+
+    Hs = H * s
+    slope = dot(g,s)
+    qs = ¹/₂sᵀHs + slope
+"""
+function compute_Hs_slope_qs!(Hs::Vector, H::Union{AbstractMatrix,AbstractLinearOperator},
+                              s::Vector, g::Vector)
+  Hs .= H * s
+  slope = dot(g,s)
+  qs = 0.5 * dot(s, Hs) + slope
+  return slope, qs
+end
+
+"""`active(x, ℓ, u; rtol = 1e-8, atol = 1e-8)`
+
+Computes the active bounds at x, using tolerance `min(rtol * (uᵢ-ℓᵢ), atol)`.
+If ℓᵢ or uᵢ is not finite, only `atol` is used.
+"""
+function active(x::Vector, ℓ::Vector, u::Vector; rtol::Real = 1e-8, atol::Real = 1e-8)
+  A = Int[]
+  n = length(x)
+  for i = 1:n
+    δ = -Inf < ℓ[i] < u[i] < Inf ? min(rtol * (u[i] - ℓ[i]), atol) : atol
+    if ℓ[i] == x[i] == u[i]
+      push!(A, i)
+    elseif x[i] <= ℓ[i] + δ
+      push!(A, i)
+    elseif x[i] >= u[i] - δ
+      push!(A, i)
+    end
+  end
+  return A
+end
+
+"""`project!(y, x, ℓ, u)`
+
+Projects `x` into bounds `ℓ` and `u`, in the sense of
+`yᵢ = max(ℓᵢ, min(xᵢ, uᵢ))`.
+"""
+function project!(y :: Vector, x :: Vector, ℓ :: Vector, u :: Vector)
+  y .= max.(ℓ, min.(x, u))
+end
+
+"""`project_step!(y, x, d, ℓ, u)`
+
+Computes the projected direction `y = P(x + d) - x`.
+"""
+function project_step!(y::Vector, x::Vector, d::Vector, ℓ::Vector, u::Vector)
+  y .= x .+ d
+  project!(y, y, ℓ, u)
+  y .-= x
+end
+
+"""`x, s = projected_line_search!(x, H, g, d, ℓ, u; μ₀ = 1e-2)`
+
+Performs a projected line search, searching for a step size `t` such that
+
+    0.5sᵀHs + sᵀg ≦ μ₀sᵀg,
+
+where `s = P(x + t * d) - x`, while remaining on the same face as `x + d`.
+Backtracking is performed from t = 1.0. `x` is updated in place.
+"""
+function projected_line_search!(x::Vector, H::Union{AbstractMatrix,AbstractLinearOperator}, g::Vector, d::Vector, ℓ::Vector, u::Vector; μ₀::Real = 1e-2)
+  α = 1.0
+  _, brkmin, _ = breakpoints(x, d, ℓ, u)
+  nsteps = 0
+  n = length(x)
+
+  s = zeros(n)
+  Hs = zeros(n)
+
+  search = true
+  while search && α > brkmin
+    nsteps += 1
+    project_step!(s, x, α * d, ℓ, u)
+    slope, qs = compute_Hs_slope_qs!(Hs, H, s, g)
+    if qs <= μ₀ * slope
+      search = false
+    else
+      α *= 0.5
+    end
+  end
+  if α < 1.0 && α < brkmin
+    α = brkmin
+    project_step!(s, x, α * d, ℓ, u)
+    slope, qs = compute_Hs_slope_qs!(Hs, H, s, g)
+  end
+
+  αd = α * d
+  project_step!(s, x, αd, ℓ, u)
+  project!(x, x .+ αd, ℓ, u)
+
+  return x, s
+end
+
+"""`α, s = cauchy(x, H, g, Δ, ℓ, u; μ₀ = 1e-2, μ₁ = 1.0, σ=10.0)`
+
+Computes a Cauchy step `s = P(x - α g) - x` for
+
+    min  q(s) = ¹/₂sᵀHs + gᵀs     s.t.    ‖s‖ ≦ μ₁Δ,  ℓ ≦ x + s ≦ u,
+
+with the sufficient decrease condition
+
+    q(s) ≦ μ₀sᵀg.
+"""
+function cauchy(x::Vector, H::Union{AbstractMatrix,AbstractLinearOperator}, g::Vector,
+                Δ::Real, α::Real, ℓ::Vector, u::Vector;
+                μ₀::Real = 1e-2, μ₁::Real = 1.0, σ::Real = 10.0)
+  # TODO: Use brkmin to care for g direction
```

Comment:
This is for when `alpha <= brkmin` and so I can reuse `H * d`. This will look ugly, and in some cases, won't improve the execution, so I think it's better to leave for a future dedicated PR.
I'll improve the TODO, though.